OpenSim/Auxiliary/auxiliaryTestFunctions.h

```diff
@@ -209,17 +209,116 @@ OpenSim::Object* randomize(OpenSim::Object* obj)
 // Storage can only read files with version <= 1.
 // This function can be removed when Storage class is removed.
 inline void revertToVersionNumber1(const std::string& filenameOld,
-                                   const std::string& filenameNew) {
-  std::regex versionline{R"([ \t]*version[ \t]*=[ \t]*\d[ \t]*)"};
-  std::ifstream fileOld{filenameOld};
-  std::ofstream fileNew{filenameNew};
-  std::string line{};
-  while(std::getline(fileOld, line)) {
-    if(std::regex_match(line, versionline))
-      fileNew << "version=1\n";
-    else
-      fileNew << line << "\n";
-  }
+    const std::string& filenameNew) {
+    std::regex versionline{ R"([ \t]*version[ \t]*=[ \t]*\d[ \t]*)" };
+    std::ifstream fileOld{ filenameOld };
+    std::ofstream fileNew{ filenameNew };
+    std::string line{};
+    while (std::getline(fileOld, line)) {
+        if (std::regex_match(line, versionline))
+            fileNew << "version=1\n";
+        else
+            fileNew << line << "\n";
+    }
+}
+
+// Estimate the memory usage of a *creator* that heap allocates an object
+// of type C and returns a pointer to it. Creator can also perform any 
+// initialization before returning the pointer.
+template <typename C, typename T>
+size_t estimateMemoryChangeForCreator(T creator, const size_t nSamples = 100)
+{
+    std::vector<std::unique_ptr<C>> pointers;
+    std::vector<size_t> deltas;
+
+    for (size_t i = 0; i < nSamples; ++i) {
+        size_t mem0 = getCurrentRSS();
+        // Execute the desired creator 
+        // store in unique_ptrs to delay deletion
+        pointers.push_back(std::unique_ptr<C>(creator()));
+        // poll the change in memory usage
+        size_t mem1 = getCurrentRSS();
+        // change in memory usage (negative values are invalid)
+        size_t delta = mem1 > mem0 ? mem1 - mem0 : 0;
+        if(delta) // store only valid values (creator cannot create 0 bytes)
+            deltas.push_back(delta);
+    }
+
+    OPENSIM_THROW_IF(deltas.size() < 2, OpenSim::Exception,
+        "Insufficient number of nonzero samples to estimate memory change. "
+        "Consider increasing the number of samples.");
+
+    size_t nmedian = deltas.size() / 2;
+    // sort the deltas up to and including the nth element
+    std::nth_element(deltas.begin(), deltas.begin() + nmedian, deltas.end());
+
+    return deltas[nmedian];
+}
+
+// Determine if getRSS is providing reliable estimates of memory usage by
+// testing against an allocation of known size and verifying that the change
+// memory use is detected. Employ this method to validate the use of
+// memory use estimators (e.g. estimateMemoryChangeForCreator and
+// estimateMemoryChangeForCommand). Do this at the beginning of your test 
+// involving checks for memory use.
+void validateMemoryUseEstimates()
+{
+    size_t nSamples = 5;
```

Comment:
> Perhaps make nSamples a default argument for symmetry with ...

Done. `nSamples` also defaults to a bigger number (11) but we would like to keep it as small as possible. Not sure if it will have an effect on OSX